ntoskrnl/ps/quota.c

```diff
@@ -18,123 +17,507 @@ EPROCESS_QUOTA_BLOCK PspDefaultQuotaBlock;
+    for (PsQuotaTypeIndex = PsQuotaTypes; PsQuotaTypeIndex <  PsQuotaTypes; PsQuotaTypeIndex++)
+    {
+        /* The amount needed to return to Mm is the limit and return fields */
+        QuotaToReturn = QuotaBlock->QuotaEntry[PsQuotaTypeIndex].Limit + QuotaBlock->QuotaEntry[PsQuotaTypeIndex].Return;
+        MmReturnPoolQuota(PsQuotaTypeIndex, QuotaToReturn);
+
+        /*
+         * If that's a page file quota then don't
+         * do anything here, so just skip it outright.
+         */
+        if (PsQuotaTypeIndex == PsPageFile)
+        {
+            break;
+        }
+    }
+}
+
+/**
+ * @brief
+ * Releases some of excess quotas in order to
+ * attempt to free up some resources. This is
+ * done primarily in case the Memory Manager
+ * fails to raise the quota limit.
+ *
+ * @param[in] QuotaType
+ * Process pool quota type.
+ *
+ * @param[out] ReturnedQuotas
+ * A pointer to the returned amount of quotas
+ * back to Memory Manager.
+ *
+ * @return
+ * Nothing.
+ *
+ * @remarks
+ * The function releases excess paged or non
+ * paged pool quotas. Page file quota type is
+ * not permitted.
+ */
+VOID
+NTAPI
+PspReturnExcessQuotas(
+    _In_ PS_QUOTA_TYPE QuotaType,
+    _Outptr_ PSIZE_T ReturnedQuotas)
+{
+    PPSP_QUOTA_LIST PspQuotaList;
+    PEPROCESS_QUOTA_BLOCK QuotaBlockFromList;
+    SIZE_T AmountToReturn = 0;
+
+    /*
+     * Loop over the quota block lists and reap
+     * whatever quotas we haven't returned which
+     * is needed to free up resources.
+     */
+    for (PspQuotaList = PspQuotaBlockList.Flink;
+         PspQuotaList != &PspQuotaBlockList;
+         PspQuotaList = PspQuotaList->Flink)
+    {
+        /* Gather that list */
+        QuotaBlockFromList = CONTAINING_RECORD(PspQuotaList, EPROCESS_QUOTA_BLOCK, QuotaList);
+
+        /*
+         * Gather any unreturned quotas and cache
+         * them to a variable. If there's nothing
+         * to return back then there's nothing to do...
+         */
+        AmountToReturn += InterlockedExchangeSizeT(&QuotaBlockFromList->QuotaEntry[QuotaType].Return, 0);
+        if (AmountToReturn == 0)
+        {
+            DPRINT1("PspReturnExcessQuotas(): No quota left to release back...\n");
+            *ReturnedQuotas = 0;
+            return;
+        }
+
+        /*
+         * If no other process is taking use of this
+         * block, then we have to check that the limit
+         * is way over the usage. If that's the case,
+         * then we'll shrink the limit up on par with
+         * the usage and the additional amount to return
+         * to the Memory Manager is that limit minus the
+         * usage.
+         */
+        if (QuotaBlockFromList->ProcessCount == 0)
+        {
+            if (QuotaBlockFromList->QuotaEntry[QuotaType].Usage <
+                QuotaBlockFromList->QuotaEntry[QuotaType].Limit)
+            {
+                InterlockedCompareExchangeSizeT(&QuotaBlockFromList->QuotaEntry[QuotaType].Limit,
```

Comment:
Is this interlocked operation just pro forma, or is there actually a chance of concurrent access here?
If the latter, then this seems like it should use fetch-once on `Usage` and `Limit`, and actually use the InterlockedCompareExchange return value. As is, we might be fetching these values 3-4 times, so if they can change in between, that wouldn't be good.

Reply:
The interlocked operation there is just pro forma, as race concurrences cannot happen here since `PspReturnExcessQuotas` is used in `PspChargeProcessQuotaSpecifiedPool` with the underlying code wrapped and guarded with a spin lock. No other threads can perform the said interlocked operation at the same time.